serialization_v2/serialization.py

```diff
@@ -0,0 +1,1280 @@
+import os
+import re
+import subprocess
+import sys
+import types
+from types import BuiltinFunctionType, FunctionType, MethodType, ModuleType
+from typing import Any, Collection, Dict, List, Optional, Sequence, Tuple, Union, TYPE_CHECKING
+
+from sisyphus import Path
+from sisyphus.delayed_ops import DelayedBase
+from sisyphus.hash import sis_hash_helper
+
+from i6_core.returnn.config import CodeWrapper, ReturnnConfig, unparse_python
+from i6_core.returnn.training import Checkpoint, PtCheckpoint
+from i6_core.serialization import (
+    Call,
+    CallImport,
+    CodeFromFile,
+    CodeFromFunction,
+    ExplicitHash,
+    ExternalImport,
+    Import,
+    NonhashedCode,
+    PartialImport,
+    SerializerObject,
+)
+from i6_core.serialization import Collection as SerializerCollection
+
+if TYPE_CHECKING:
+    from returnn.tensor import Dim
+    from returnn.util.file_cache import CachedFile
+
+
+def serialize_config(
+    config: Dict[str, Any],
+    post_config: Optional[Dict[str, Any]] = None,
+    *,
+    inlining: bool = True,
+    known_modules: Collection[str] = (),
+    extra_sys_paths: Sequence[str] = (),
+    sis_path_handling: SisPathHandling = None,
+) -> SerializedConfig:
+    """serialize config. see module docstring for more info."""
+    serializer = _Serializer(
+        config=config, post_config=post_config, known_modules=known_modules, sis_path_handling=sis_path_handling
+    )
+    for path in extra_sys_paths:
+        serializer.add_sys_path(path, recursive=False)
+    serializer.work_queue()
+    if inlining:
+        serializer.work_inlining()
+    return SerializedConfig(code_list=list(serializer.assignments_dict_by_idx.values()))
+
+
+class SisPathHandling(enum.Enum):
+    """
+    SisPathHandling enum.
+    """
+
+    NONE = None
+    AS_STRING = "as_string"
+    NO_DEPS = "no_deps"
+
+
+def _instanciate_delayed_copy(o: Any) -> Any:
+    """
+    Recursively traverses a structure and calls .get() on all
+    existing Delayed Operations, especially Variables in the structure
+
+    In contrast to :func:`i6_core.util.instanciate_delayed` this function does not operate inplace.
+
+    :param o: nested structure that may contain DelayedBase objects
+    :return: o with all DelayedBase objects replaced by their .get() value
+    """
+    import tree
+
+    def _instanciate_delayed_obj(o: Any) -> Any:
+        if isinstance(o, DelayedBase) and not isinstance(o, SerializerObject):
+            return o.get()
+        return o
+
+    return tree.map_structure(_instanciate_delayed_obj, o)
+
+
+def _is_valid_python_identifier_name(name: str) -> bool:
+    """
+    :return: whether the name is a valid Python identifier name (including attrib name)
+    """
+    return name.isidentifier() and not iskeyword(name)
+
+
+class ReturnnConfigWithNewSerialization(ReturnnConfig):
+    """
+    Overwrites the serialization behavior of ReturnnConfig.
+    """
+
+    @staticmethod
+    def from_cfg(old_returnn_cfg: ReturnnConfig):
```

Comment:
I think this cannot be done in a safe way.

Reply:
I thought mostly about using Python epilog / prolog when making this comment. As you want to keep this, and you accept that this is not safe, then also this function here could stay.

But I would say, only if you really use it. Do you?

Reply:
Well, I mean, I did not suggest to remove it. I just wanted to understand the use case. But ok, if you think it's better to remove, then let's keep it removed for now.